x-pack/metricbeat/module/azure/billing/client.go

```diff
@@ -42,55 +44,74 @@ func NewClient(config azure.Config) (*Client, error) {
 }
 
 // GetMetrics returns the usage detail and forecast values.
-func (client *Client) GetMetrics(startTime time.Time, endTime time.Time) (Usage, error) {
+func (client *Client) GetMetrics(timeOpts TimeIntervalOptions) (Usage, error) {
 	var usage Usage
+
+	//
+	// Establish the requested scope
+	//
+
 	scope := fmt.Sprintf("subscriptions/%s", client.Config.SubscriptionId)
 	if client.Config.BillingScopeDepartment != "" {
 		scope = fmt.Sprintf("/providers/Microsoft.Billing/departments/%s", client.Config.BillingScopeDepartment)
 	} else if client.Config.BillingScopeAccountId != "" {
 		scope = fmt.Sprintf("/providers/Microsoft.Billing/billingAccounts/%s", client.Config.BillingScopeAccountId)
 	}
+
+	//
+	// Fetch the usage details
+	//
+
 	client.Log.
 		With("billing.scope", scope).
-		With("billing.start_time", startTime).
-		With("billing.end_time", endTime).
+		With("billing.usage.start_time", timeOpts.usageStart).
+		With("billing.usage.end_time", timeOpts.usageEnd).
 		Infow("Getting usage details for scope")
 
-	usageDetails, err := client.BillingService.GetUsageDetails(
+	filter := fmt.Sprintf(
+		"properties/usageStart eq '%s' and properties/usageEnd eq '%s'",
+		timeOpts.usageStart.Format(time.RFC3339Nano),
+		timeOpts.usageEnd.Format(time.RFC3339Nano),
+	)
+
+	paginator, err := client.BillingService.GetUsageDetails(
 		scope,
 		"properties/meterDetails",
-		fmt.Sprintf(
-			"properties/usageStart eq '%s' and properties/usageEnd eq '%s'",
-			startTime.Format(time.RFC3339Nano),
-			endTime.Format(time.RFC3339Nano),
-		),
+		filter,
 		"", // skipToken
+		//&pageSize,
 		nil,
 		consumption.MetrictypeActualCostMetricType,
-		startTime.Format("2006-01-02"), // startDate
-		endTime.Format("2006-01-02"),   // endDate
+		timeOpts.usageStart.Format("2006-01-02"), // startDate
+		timeOpts.usageEnd.Format("2006-01-02"),   // endDate
 	)
 	if err != nil {
 		return usage, fmt.Errorf("retrieving usage details failed in client: %w", err)
 	}
 
-	usage.UsageDetails = usageDetails.Values()
+	for paginator.NotDone() {
```

Comment:
The module collects Usage Details data and forecast data at different time intervals:

* usage details: previous day (24 hours)
* forecast: current month (30 days)

Suppose we have zero usage in the last 24h and some resource usage in the previous days. In this case, usage details would be empty, but we could have some forecasts based on the days before.

WDYT? I have no previous knowledge of this topic, I reconstructed it from the codebase and Microsoft docs, so any question like this one is more than welcome!

Reply:
> Suppose we have zero usage in the last 24h and some resource usage in the previous days. In this case, usage details would be empty, but we could have some forecasts based on the days before.
> 
> WDYT? I have no previous knowledge of this topic, I reconstructed it from the codebase and Microsoft docs, so any question like this one is more than welcome!

I indeed didn't think about the fact that the call to the usage details could fail because there are zero usage in the last 24h :)

Mine was a more generic doubt about the fact that they are two different API calls, that can fail each for different reasons, even transient (ie: by the time we call the first we have a network outage that will be solved by the time of the next call, or just one upstream service is down while the other is up). Does it make sense to keep fetching the second metrics if fetching the first fails?

I would look the answer to this in the dashboards or the way the two metrics are related to each others.
Taking the example you made, and assuming having zero usage data is the reason for a possible failure: I see it can make sense to see in a dashboard that I have zero usage and the forecast is horizontal, rather than missing at all any information. At the same time is the usage is zero and the forecast increase this can be an hint that something is not working properly.

Probably @ravikesarwani might have an opinion here as well.

Reply:
> Does it make sense to keep fetching the second metrics if fetching the first fails?

Yep, I think it makes sense.

In the dashboard, usage details and forecasts have independent visualizations: we don't have one visualization that uses data from both. However, the sum of the usage details matches the amount of actual data in the forecast for the same day.

Reply:
I'm in favor of changing the logic and continuing with the forecast even if the call for usage details fails for some reason. 

@ravikesarwani, let us know what's on your mind.